service/history/queues/executable.go

```diff
@@ -334,62 +312,112 @@ func (e *executableImpl) HandleErr(err error) (retErr error) {
 			}
 			e.resourceExhaustedCount++
 			metrics.TaskThrottledCounter.With(e.taggedMetricsHandler).Record(1)
-			return err
+			return true, err
 		}
 
 		err = consts.ErrResourceExhaustedBusyWorkflow
 	}
 	e.resourceExhaustedCount = 0
 
 	if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
-		return nil
+		return true, nil
 	}
 
 	// This means that namespace is deleted, and it is safe to drop the task (=ignore the error).
 	if _, isNotFound := err.(*serviceerror.NamespaceNotFound); isNotFound {
-		return nil
+		return true, nil
+	}
+
+	if _, isDataLoss := err.(*serviceerror.DataLoss); isDataLoss {
+		e.logger.Error("Dropping task due to data loss", tag.Error(err))
+		return true, nil
 	}
 
+	if _, ok := err.(*serviceerror.NamespaceNotActive); ok {
+		// error is expected when there's namespace failover,
+		// so don't count it into task failures.
+		metrics.TaskNotActiveCounter.With(e.taggedMetricsHandler).Record(1)
+		return true, err
+	}
+
+	return false, err
+}
+
+// Terminal history errors are those we've defined as neither
+// retryable nor worth sending to the DLQ.
+// TODO: We need a better name for this. ImpossibleToFixError? IsTerminalButNotDLQableError?
+func (e *executableImpl) isTerminalHistoryError(err error) (bool, error) {
```

Comment:
I am able to understand the logic but I feel this is very confusing.

The approach in my mind is:
- A function called `isSafeToDropError` which checks for those errors that are converted to nil. e.g. NotFound or ErrTaskVersionMismatch. This is expected non-retryable errors I'd say.
- A another function  `isExpectedRetryableError` which checks for errors like ErrDependencyTaskNotCompleted , ErrTaskRetry, or NamespaceNotActive
- Then the reset are unexpected errors. We can log the "Fail to process task" log and emit TaskFailure metric.
- We can define `isUnexpectedNonRetryableError` which put task into DLQ if enabled or drop the task. Within the function we can check for is the errors implements the TerminalTaskError interface.
- The last category is unexpected retryable errors and there's nothing special we need to for them.

Reply:
I like this way of breaking up the error classes